app/presenters/two_factor_add_methods_presenter.rb

```diff
@@ -0,0 +1,65 @@
+class TwoFactorAddMethodsPresenter
+  include ActionView::Helpers::TranslationHelper
+
+  attr_reader :current_user, :service_provider, :next_step_url
+
+  def initialize(current_user, sp, view, next_step_url)
+    @current_user = current_user
+    @service_provider = sp
+    @next_step_url = next_step_url
+    @view = view
+  end
+
+  def title
+    t('devise.two_factor_authentication.two_factor_login_add_methods_title')
+  end
+
+  def heading
+    t('devise.two_factor_authentication.two_factor_login_add_methods_title')
+  end
+
+  def info
+    t('devise.two_factor_authentication.two_factor_login_add_methods_intro')
+  end
+
+  def skip_step_link
+    @view.link_to(t('devise.two_factor_authentication.add_methods_later'), @next_step_url)
+  end
+
+  def label
+    ''
+  end
+
+  def options
+    available_2fa_types.map do |type|
+      OpenStruct.new(
+        type: type,
+        label: t("devise.two_factor_authentication.two_factor_login_options.#{type}"),
+        info: t("devise.two_factor_authentication.two_factor_login_options.#{type}_info"),
+        selected: type == :auth_app
+      )
+    end
+  end
+
+  private
+
+  def available_2fa_types
+    phone_if_available + auth_app_if_available + piv_cac_if_available
+  end
+
+  def auth_app_if_available
+    return [] if current_user.totp_enabled?
+    %w[auth_app]
+  end
+
+  def phone_if_available
+    return [] if current_user.phone_enabled?
+    %w[sms voice]
+  end
+
+  def piv_cac_if_available
+    return [] if current_user.piv_cac_enabled?
+    return [] unless current_user.piv_cac_available? || service_provider&.piv_cac_available?
+    %w[piv_cac]
+  end
```

Comment:
What do you think about reducing this complexity by adding a new Postgres array field to the Users table that will hold all the 2FA options they have set up so far? That way, all you need is a diff between what they already have and what's available (which should be a constant) as opposed to asking if they are phone_enabled, totp_enabled, etc.

Reply:
This was going to be part of @jgsmith-usds 's refactor.  The logic was going to be extracted out into a service.

Reply:
It definitely needs work.  If we were to store it in the db I'd opt for exploding it out into a table.  I'm not a huge fan of array fields.

Reply:
I'd be happy with getting this logic out of the presenter as a first step.

Reply:
I'd rather not store it in an array when we can derive the information from the configuration. It's begging to get out of sync. Besides, I'm trying to make some kind of progress on refactoring the 2fa support into a modular system. I don't want to make things any more confusing than they already are.

Reply:
What will get out sync? The array that I'm proposing is for storing the options that the user has configured so far, not the total possible options, which I would assume would be defined as a constant.

Now that I see what feature we want to implement, I think the design we have works fine. We got rid of a lot of the complexity in generating the links to the alternative options, which is great. I love the single link that points to a dedicated page that displays which options are available.

Each option points to a specific controller, which points to a specific Form Object for setting up that option. What isn't working with this design?

The only remaining complexity is the logic that determines which options to display, and I think that storing the options that the user has configured so far in an array would be the least complex solution. It would look something like this:
```ruby
POSSIBLE_OPTIONS = %w[sms voice auth_app piv_cac].freeze # this will only change when we add new options

already_configured_options = current_user.configured_2fa_options
available_options = POSSIBLE_OPTIONS - already_configured_options
rule = UserEligibleForPivCacRule.new(user: user, service_provider: service_provider)
available_options << 'piv_cac' if rule.satisfied?

class UserEligibleForPivCacRule
  def initialize(user:, service_provider:)
    @user = user
    @service_provider = service_provider
  end
  
  def satisfied?
    return false if user_already_has_piv_cac?
    user.piv_cac_available? || service_provider&.piv_cac_available?
  end
  
  private

  def user_already_has_piv_cac?
    user.configured_2fa_options.include?('piv_cac')
  end
end
```

Reply:
This code looks great.  When you mentioned array fields I thought you meant bit fields which are crappy to work with. The array fields in Postgres look pretty cool.  I didn't realize that you could query them with straight SQL although it's non-standard SQL.

Reply:
I would still prefer full blown tables with one table defining the 2fa types and their capabilities and another table as a one to many on users specifying which types the user has.

Reply:
Modifying the database and refactoring our 2fa system is out of scope for this PR.